File: src/ripple/agent_runners/approvals.py
"""Codex app-server approval request helpers."""

from typing import Any, Literal

ApprovalAction = Literal["allow", "always", "deny"]

_APPROVAL_METHOD_ACTIONS = {
    "item/commandExecution/requestApproval": "command_execution",
    "item/fileChange/requestApproval": "file_change",
    "item/permissions/requestApproval": "permissions",
    "execCommandApproval": "exec_command",
    "applyPatchApproval": "apply_patch",
}


def parse_codex_approval_request(
    message: dict[str, Any],
    *,
    job_id: str,
    user_id: str | None,
    session_id: str | None,
) -> dict[str, Any] | None:
    method = message.get("method")
    if not isinstance(method, str) or method not in _APPROVAL_METHOD_ACTIONS:
        return None
    request_id = message.get("id")
    if request_id is None:
        return None

    params = message.get("params")
    params = params if isinstance(params, dict) else {}
    action = _APPROVAL_METHOD_ACTIONS[method]
    description = _approval_description(action, params)

    return {
        "source": "codex",
        "job_id": job_id,
        "user_id": user_id,
        "session_id": session_id,
        "thread_id": params.get("threadId") or params.get("conversationId"),
        "turn_id": params.get("turnId"),
        "request_id": request_id,
        "method": method,
        "action": action,
        "description": description,
        "metadata": params,
    }


def codex_approval_response_for_action(approval: dict[str, Any], action: ApprovalAction) -> dict[str, Any]:
    approval_action = approval.get("action")
    if approval_action in {"command_execution", "exec_command"}:
        if action == "allow":
            return {"decision": "accept"} if approval_action == "command_execution" else {"decision": "approved"}
        if action == "always":
            return (
                {"decision": "acceptForSession"}
                if approval_action == "command_execution"
                else {"decision": "approved_for_session"}
            )
        return {"decision": "decline"} if approval_action == "command_execution" else {"decision": "denied"}

    if approval_action in {"file_change", "apply_patch"}:
        if action == "allow":
            return {"decision": "accept"} if approval_action == "file_change" else {"decision": "approved"}
        if action == "always":
            return {"decision": "acceptForSession"} if approval_action == "file_change" else {"decision": "approved"}
        return {"decision": "decline"} if approval_action == "file_change" else {"decision": "denied"}

    if approval_action == "permissions":
        if action == "deny":
            return {"permissions": {}, "scope": "turn", "strictAutoReview": True}
        metadata = approval.get("metadata")
        metadata = metadata if isinstance(metadata, dict) else {}
        permissions = metadata.get("permissions")
        permissions = permissions if isinstance(permissions, dict) else {}
        return {
            "permissions": permissions,
            "scope": "session" if action == "always" else "turn",
            "strictAutoReview": False,
        }

    if action == "deny":
        return {"decision": "decline"}
    return {"decision": "acceptForSession" if action == "always" else "accept"}


def _approval_description(action: str, params: dict[str, Any]) -> str:
    if action in {"command_execution", "exec_command"}:
        command = params.get("command")
        if isinstance(command, list):
            return " ".join(str(part) for part in command)
        if isinstance(command, str) and command:
            return command
    if action == "file_change":
        grant_root = params.get("grantRoot")
        if isinstance(grant_root, str) and grant_root:
            return f"Allow file changes under {grant_root}"
    if action == "permissions":
        reason = params.get("reason")
        if isinstance(reason, str) and reason:
            return reason
    reason = params.get("reason")
    if isinstance(reason, str) and reason:
        return reason
    return action.replace("_", " ")

File: crates/ripple-server/src/codex/approvals.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodexApproval {
    pub source: String,
    pub job_id: String,
    pub user_id: Option<String>,
    pub session_id: Option<String>,
    pub request_id: Value,
    pub method: String,
    pub action: String,
    pub description: String,
    pub metadata: Value,
}

pub fn approval_response_for_action(approval: &CodexApproval, action: &str) -> Value {
    match approval.action.as_str() {
        "command_execution" | "exec_command" => {
            if action == "deny" {
                if approval.action == "exec_command" {
                    serde_json::json!({"decision": "denied"})
                } else {
                    serde_json::json!({"decision": "decline"})
                }
            } else if action == "always" {
                if approval.action == "exec_command" {
                    serde_json::json!({"decision": "approved_for_session"})
                } else {
                    serde_json::json!({"decision": "acceptForSession"})
                }
            } else if approval.action == "exec_command" {
                serde_json::json!({"decision": "approved"})
            } else {
                serde_json::json!({"decision": "accept"})
            }
        }
        "file_change" | "apply_patch" => {
            if action == "deny" {
                if approval.action == "apply_patch" {
                    serde_json::json!({"decision": "denied"})
                } else {
                    serde_json::json!({"decision": "decline"})
                }
            } else if action == "always" && approval.action == "file_change" {
                serde_json::json!({"decision": "acceptForSession"})
            } else if approval.action == "apply_patch" {
                serde_json::json!({"decision": "approved"})
            } else {
                serde_json::json!({"decision": "accept"})
            }
        }
        "permissions" => {
            if action == "deny" {
                serde_json::json!({"permissions": {}, "scope": "turn", "strictAutoReview": true})
            } else {
                let permissions = approval
                    .metadata
                    .get("permissions")
                    .cloned()
                    .unwrap_or_else(|| serde_json::json!({}));
                serde_json::json!({
                    "permissions": permissions,
                    "scope": if action == "always" { "session" } else { "turn" },
                    "strictAutoReview": false
                })
            }
        }
        _ => {
            if action == "deny" {
                serde_json::json!({"decision": "decline"})
            } else {
                serde_json::json!({"decision": if action == "always" { "acceptForSession" } else { "accept" }})
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{approval_response_for_action, CodexApproval};

    fn permissions_approval() -> CodexApproval {
        CodexApproval {
            source: "codex_app_server".to_string(),
            job_id: "job-1".to_string(),
            user_id: Some("alice".to_string()),
            session_id: Some("session-1".to_string()),
            request_id: json!("request-1"),
            method: "item/permissions/requestApproval".to_string(),
            action: "permissions".to_string(),
            description: "Needs folder access".to_string(),
            metadata: json!({
                "reason": "Read sibling folder",
                "cwd": "/workspace/spaces/a/records/r1",
                "permissions": {
                    "filesystem": {
                        "read": ["/workspace/spaces/a/records/r2"],
                        "write": ["/workspace/spaces/a/records/r2/out.md"]
                    }
                }
            }),
        }
    }

    #[test]
    fn permission_allow_returns_requested_permissions_only() {
        let approval = permissions_approval();
        let response = approval_response_for_action(&approval, "allow");

        assert_eq!(
            response.get("permissions"),
            approval.metadata.get("permissions")
        );
        assert_eq!(response.get("scope"), Some(&json!("turn")));
        assert_eq!(response.get("strictAutoReview"), Some(&json!(false)));
    }

    #[test]
    fn permission_always_maps_to_session_scope() {
        let approval = permissions_approval();
        let response = approval_response_for_action(&approval, "always");

        assert_eq!(
            response.get("permissions"),
            approval.metadata.get("permissions")
        );
        assert_eq!(response.get("scope"), Some(&json!("session")));
    }

    #[test]
    fn permission_deny_returns_empty_permissions() {
        let approval = permissions_approval();
        let response = approval_response_for_action(&approval, "deny");

        assert_eq!(response.get("permissions"), Some(&json!({})));
        assert_eq!(response.get("scope"), Some(&json!("turn")));
        assert_eq!(response.get("strictAutoReview"), Some(&json!(true)));
    }
}
